File: src/main/java/com/meepoffaith/hextrapats/init/ContinuationTypes.java
package com.meepoffaith.hextrapats.init;

import at.petrak.hexcasting.api.casting.eval.vm.ContinuationFrame;
import at.petrak.hexcasting.xplat.IXplatAbstractions;
import com.meepoffaith.hextrapats.HextraPatterns;
import com.meepoffaith.hextrapats.casting.eval.vm.FrameIndexForEach;
import com.meepoffaith.hextrapats.casting.eval.vm.FrameMainForEach;
import net.minecraft.registry.Registry;

public class ContinuationTypes{
    public static final ContinuationFrame.Type<FrameIndexForEach> INDEX_FOREACH = continuation("index_foreach", FrameIndexForEach.TYPE);
    public static final ContinuationFrame.Type<FrameMainForEach> MAIN_FOREACH = continuation("main_foreach", FrameMainForEach.TYPE);

    public static void init(){}

    private static <U extends ContinuationFrame, T extends ContinuationFrame.Type<U>> T continuation(String name, T continuation){
        return Registry.register(IXplatAbstractions.INSTANCE.getContinuationTypeRegistry(), HextraPatterns.modLoc(name), continuation);
    }
}

File: src/main/kotlin/com/meepoffaith/hextrapats/init/ContinuationTypes.kt
package com.meepoffaith.hextrapats.init

import at.petrak.hexcasting.api.casting.eval.vm.ContinuationFrame
import at.petrak.hexcasting.xplat.IXplatAbstractions
import com.meepoffaith.hextrapats.HextraPats
import com.meepoffaith.hextrapats.casting.eval.vm.FrameForEachIndex
import com.meepoffaith.hextrapats.casting.eval.vm.FrameMainForEach
import net.minecraft.registry.Registry

object ContinuationTypes {
    val FOREACH_INDEX = registerContinuation("foreach_index", FrameForEachIndex.TYPE)
    val MAIN_FOREACH = registerContinuation("main_foreach", FrameMainForEach.TYPE)

    fun init(){}

    private fun registerContinuation(name: String, continuation: ContinuationFrame.Type<*>) :  ContinuationFrame.Type<*>{
        return Registry.register(IXplatAbstractions.INSTANCE.continuationTypeRegistry, HextraPats.modLoc(name), continuation)
    }
}
